README.md

Tseries - Meteorological Timeseries Viewer
==========================================

Norwegian Meteorological Institute (met.no)  
Box 43 Blindern  
0313 OSLO  
NORWAY  
email: diana@met.no

About Tseries
----------------

Tseries is part of the Diana package from met.no (Norwegian Meteorological Institute) and is distributed under the GPL license. See gpl.txt for details concerning the GPL license.

Tseries is a graphical time series viewer developed for use with meteorological and oceanographic data. It uses OpenGL for the graphics and Trolltech Qt for the user interface.

Tseries supports multiple languages via the Qt linguist system. Additional languages can be added by preparing new `lang/tseries_XX.ts` files (using the Qt linguist program) and adding the new language to the build process (`src/CMakeLists.txt`). 

Depends on:
 * [Qt4](http://qi.io) for the user interface and the drawing code
 * [HDF 4.2](http://www.hdfgroup.org/)
 * [fimex](http://fimex.met.no)
 * MET Norway [metlibs](https://svn.met.no/viewvc/metlibs/)
 * [CMake](https://cmake.org/) (above version 2.8.7)

Installation
------------

For ubuntu users, Tseries is available from the [MET Norway launchpad PPA](https://launchpad.net/~met-norway/+archive/ubuntu/metapps).

To compile Tseries yourself, you first need to compile the required fimex and metlibs libraries and install them. Then you may use something like this to compile Tseries:

    export PACKAGE_CONFIG_PATH=where_you_installed_metlibs/lib/pkgconfig
    BUILD=/path/to/tseries/build
    SOURCE=/path/to/tseries/source
    mkdir -p $BUILD
    cd $BUILD
    cmake $SOURCE
    make
    make install

Running Tseries
---------------

running Tseries with the prepared test datasets and configurations:

1. The test data sets should be installed under `metnolocal/demodata/`
2. Starting Tseries
        cd metnolocal/tseries
        bin/tseries -s tseries.ctl_export

Select the language of preference under "Preferences | languages" in the menu.
User guide available under Help.

Configuration files
--------------------

`tseries.ctl_export`:
  The setup file for Tseries. 
  Change top level directory for installation with `Topdir` variable
  Add new datafiles under section `<streams> `

`etc/tsDiagrams_export.def`:
  Definition of diagram types with supported data
  See comments in file for syntax
  IMPORTANT: if you change the top level directory in `tseries.ctl_export`,
  you must also update the style paths here (see `STYLEFILE=`).

`.tseries.conf`:
  User preferences saved here.

`etc/tseries.filter`:
  User defined position filter saved here (see online doc)

`etc/symbols.def`:
  Weather symbol mapping

`style/style.*`
  Detailed instructions for the plotting package (in ascii). 
  Determines which components to add to a specific diagram.
  Used in `etc/tsDiagrams_export.def`. The structure of these files
  is currently undocumented.

File formats
------------

Tseries supports a number of in-house (met.no) formats. 

1. The most commonly used is the HDF 4.2 based format. In the <streams> section of the setup file you will find references to files of this type with "DataType=HDF". 
The exact specification for this format is omitted here; mainly because we intend to make changes to this format in the near future. If you are still interested: use the 'vshow' program that comes with the HDF4.2 package to view the structure of one the files in the test data set.

2. Tseries also supports a simple ascii format. See the test data file `metnolocal/demodata/hirlam/t2m_lqr.txt` for a complete example. 
Here is a short example:

        # --------------------------------------
        # Example time series data file in ascii
        # --------------------------------------
        
        #- source name, runtime
        Model=HIRLAM20, 2006-05-24 00:00:00
        
        #-----------------
        # New position
        #-----------------
        #- Name of position, longitude, latitude
        Position=GARDERMOEN, 11.1, 60.2
        
        #- Data-header
        #  the Time and Level specification is mandatory
        #  Add your own parameters after these (defined properly in etc/parameters.def)
        Time                  Level     MSLP    TT
        #- Data-table matching header
        2006-05-23 18:00:00   0         1000.1  23.3
        2006-05-24 00:00:00   0         999.9   16.4
        .....
        
        #-----------------
        # Start a new position with Position=
        #-----------------
        #- Name of position, longitude, latitude
        Position=X,....
        ....
        # End of Ascii file format description
# tseries-qt4


debian_files/changelog
tseries (7.0.6-1) unstable; urgency=low

  * optional disabling mora streams / fixing timeslider
  * do not work with undefined miTime
  * added RRAC_x streams to enable MEPS

 -- Juergen Schulze <juergen.schulze@met.no>  Tue, 09 Aug 2016 13:13:09 +0200

tseries (7.0.5-1) unstable; urgency=low

  * bugfix show/hide icons in diana

 -- Alexander Bürger <alexander.buerger@met.no>  Wed, 29 Jun 2016 08:00:00 +0200

tseries (7.0.4-1) unstable; urgency=low

  * added man page
  * adapt for metlibs-coserver 2.1.0

 -- Alexander Bürger <alexander.buerger@met.no>  Thu, 26 May 2016 08:00:00 +0200

tseries (7.0.3-1) unstable; urgency=low

  * print in landscape by default
  * rename tseries-X.Y.bin to tseries.bin-X.Y

 -- Alexander Bürger <alexander.buerger@met.no>  Fri, 05 May 2016 08:00:00 +0200

tseries (7.0.2-1) unstable; urgency=low

  * removed some old print support classes
  * bugfixes CMakeLists.txt
  * fewer compilation warnings
  * require newest pets2

 -- Alexander Bürger <alexander.buerger@met.no>  Tue, 03 May 2016 08:00:00 +0200

tseries (7.0.1-2) unstable; urgency=low

  * removed update-alternatives invocation

 -- Alexander Bürger <alexander.buerger@met.no>  Mon, 02 May 2016 08:00:00 +0200

tseries (7.0.1-1) unstable; urgency=low

  * removed MET Norway configuration

 -- Alexander Bürger <alexander.buerger@met.no>  Mon, 02 May 2016 08:00:00 +0200

tseries (7.0.0-1) unstable; urgency=low

  * use pets2 6.0.0 (without OpenGL)
  * CMake build

 -- Alexander Bürger <alexander.buerger@met.no>  Thu, 28 Apr 2016 08:00:00 +0200

tseries (6.9.3-1) unstable; urgency=low

  * removed setDirectRendering
  * moved print-initialising from start to print() to avoid startup delay

 -- Juergen Schulze <juergen.schulze@met.no>  Fri, 19 Feb 2016 09:14:23 +0100

tseries (6.9.2-1) unstable; urgency=low

  * added test for lustre-disk

 -- Juergen Schulze <juergen.schulze@met.no>  Tue, 09 Feb 2016 13:08:19 +0100

tseries (6.9.1-1) unstable; urgency=low

  * added lustreCheck to script

 -- Juergen Schulze <juergen.schulze@met.no>  Tue, 15 Dec 2015 12:52:43 +0100

tseries (6.9.0-2) UNRELEASED; urgency=medium

  * added start option for VGL

 -- Juergen Schulze <juergen.schulze@met.no>  Fri, 30 Oct 2015 08:44:37 +0000

tseries (6.9.0-1) unstable; urgency=low

  * Added Mywave-wam datastream
  * Fixed a bug in reading wave and ocean models
  * Added setup option to disable datasources from tabs (disabled deprecated WDB tab)

 -- Juergen Schulze <juergen.schulze@met.no>  Mon, 19 Oct 2015 09:15:53 +0200

tseries (6.8.1-1) unstable; urgency=low

  * Fix in case of undef data (fimex)
  * Changed format for runtimes to ensure correct sorting

 -- Juergen Schulze <juergen.schulze@met.no>  Fri, 22 May 2015 12:12:30 +0200

tseries (6.8.0-2) unstable; urgency=low

  * New build with new tsData for ensemble precipitation support (fimex)

 -- Juergen Schulze <juergen.schulze@met.no>  Wed, 04 Mar 2015 13:42:47 +0100

tseries (6.8.0-1) unstable; urgency=low

  * added multidimensional arrays for eps processing in fimex stream

 -- Juergen Schulze <juergen.schulze@met.no>  Fri, 27 Feb 2015 13:39:08 +0100

tseries (6.7.0-1) unstable; urgency=low

  * Added search in halo placedatabase
  * Added target button in fimexdialog

 -- Juergen Schulze <juergen.schulze@met.no>  Fri, 12 Dec 2014 13:13:09 +0100

tseries (6.6.0-1) unstable; urgency=low

  * Adaption to search filter - setCurrent to first item in filter

 -- Juergen Schulze <juergen.schulze@met.no>  Tue, 21 Oct 2014 12:01:59 +0200

tseries (6.5.3-1) unstable; urgency=low

  * GFS timeline quick fix (test for basetimeline)
  * LC_NUMERIC=C due to problems at VV

 -- Juergen Schulze <juergen.schulze@met.no>  Thu, 26 Jun 2014 10:26:53 +0200

tseries (6.5.2-1) unstable; urgency=low

  * removed LC_NUMERIC

 -- Juergen Schulze <juergen.schulze@met.no>  Tue, 24 Jun 2014 13:48:55 +0200

tseries (6.5.1-1) unstable; urgency=low

  * fixed fimextab cut and paste

 -- Juergen Schulze <juergen.schulze@met.no>  Tue, 24 Jun 2014 10:58:35 +0200

tseries (6.5.0-1) unstable; urgency=low

  * Changed fimextab
  * added a stationfilter

 -- Juergen Schulze <juergen.schulze@met.no>  Tue, 24 Jun 2014 07:12:59 +0200

tseries (6.4.2-2) unstable; urgency=low

  * build with fimex 0.51

 -- Juergen Schulze <juergen.schulze@met.no>  Tue, 25 Feb 2014 10:16:23 +0100

tseries (6.4.2-1) unstable; urgency=low

  * Patch to put the fimex configure into the FimexStream constructor 

 -- Juergen Schulze <juergen.schulze@met.no>  Wed, 22 Jan 2014 13:50:04 +0100

tseries (6.4.1-1) unstable; urgency=low

  * Added degree of freedom to fimex input in tseries.ctl (parametertype)

 -- Juergen Schulze <juergen.schulze@met.no>  Tue, 21 Jan 2014 13:13:29 +0100

tseries (6.4.0-1) unstable; urgency=low

  * Documentation moved to the web
  * Added feature - fimex configfile
  * Added feature - nowtime as formatted string available in setup

 -- Juergen Schulze <juergen.schulze@met.no>  Fri, 17 Jan 2014 13:31:50 +0100

tseries (6.3.1-1) unstable; urgency=low

  * Fixed problem with missing runtimes in station  modus

 -- Juergen Schulze <juergen.schulze@met.no>  Mon, 16 Dec 2013 09:40:37 +0100

tseries (6.3.0-1) unstable; urgency=low

  * enabled multithreaded reading of fimex streams
  * progress bar
  * added PVERSION to path for standard setup search
  * cached Klima query

 -- Juergen Schulze <juergen.schulze@met.no>  Thu, 05 Dec 2013 12:32:57 +0100

tseries (6.2.3-1) unstable; urgency=low

  * added function to use filename instead of runtime in fimex (for files without runtime)

 -- Juergen Schulze <juergen.schulze@met.no>  Mon, 25 Nov 2013 08:35:50 +0100

tseries (6.2.2-1) unstable; urgency=low

  * fixed small problem under config read for fimex

 -- Juergen Schulze <juergen.schulze@met.no>  Tue, 19 Nov 2013 14:14:08 +0100

tseries (6.2.1-1) unstable; urgency=low

  * fixed problem with quba connections

 -- Juergen Schulze <juergen.schulze@met.no>  Mon, 18 Nov 2013 13:38:21 +0100


src/tseries.cc
/*
  Tseries - A Free Meteorological Timeseries Viewer

  Copyright (C) 2006-2016 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of Tseries

  Tseries is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  Tseries is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Tseries; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */


#include "tsSetup.h"
#include "tsConfigure.h"
#include "qtsMain.h"

#include <puTools/miCommandLine.h>
#include <puTools/miTime.h>
#include <puTools/miStringFunctions.h>

#include <QTranslator>
#include <QApplication>

#include <fstream>

using namespace std;
using namespace miutil;

int main(int argc, char **argv)
{
  vector<miCommandLine::option> o;
  o.push_back(miCommandLine::option( 's',"setup"    , 1 ));
  o.push_back(miCommandLine::option( 'h',"help"     , 0 ));
  o.push_back(miCommandLine::option( 'f',"file"     , 1 ));
  o.push_back(miCommandLine::option( 'S',"site"     , 1 ));
  o.push_back(miCommandLine::option( 'l',"lang"     , 1 ));
  o.push_back(miCommandLine::option( 'T',"title"    , 1 ));
  o.push_back(miCommandLine::option( 'I',"instancename", 1 ));
  o.push_back(miCommandLine::option( 'd',"define"   , 1 ));
  o.push_back(miCommandLine::option( 'H',"host"     , 1 ));
  o.push_back(miCommandLine::option( 'u',"user"     , 1 ));

  miCommandLine cl(o,argc, argv);

  // read setup

  string    site         = "FOU";
  string    setupfile    = "tseries.ctl";
  string    title        = "T-series";
  string    instancename = "";
  tsSetup     setup;
  tsConfigure config;
  std::string    lang;

  if(cl.hasFlag('h')){
    cerr << "Usage: "
        << argv[0]
                << "  -s setupfile "  << endl
                << "  -S site      "  << endl
                << "  -l lang      "  << endl
                << "  -T title     "  << endl
                << "  -I instancename" << endl
                << "  -H wdbhost   "  << endl
                << "  -u wdbuser   "  << endl
                << "  -d section1:key1=token1 section2:key2=token2 "
                << endl << endl;
    exit (0);
  }

  if(cl.hasFlag('S')) site         = cl.arg('S')[0];
  if(cl.hasFlag('s')) setupfile    = cl.arg('s')[0];
  if(cl.hasFlag('T')) title += " " + cl.arg('T')[0];
  if(cl.hasFlag('I')) instancename = (cl.arg('I')[0]);


  if(!setup.read(setupfile,site))
    exit(0);



  if(!setup.lang.empty())
    lang=setup.lang;

  config.read(setup.files.configure,instancename);
  config.get("LANG",lang);

  if(cl.hasFlag('d')) {
    vector<std::string> overridetokens=cl.arg('d');
    for (vector<std::string>::iterator it = overridetokens.begin(); it != overridetokens.end(); ++it)
      setup.overrideToken(*it);
  }

  if(cl.hasFlag('H')) setup.overrideToken("WDB:host="+cl.arg('H')[0]);
  if(cl.hasFlag('u')) setup.overrideToken("WDB:user="+cl.arg('u')[0]);


  if(cl.hasFlag('l')) lang=cl.arg('l')[0];


  QApplication a( argc, argv );
  QTranslator  myapp( 0 );
  QTranslator  qt( 0 );

  miTime defTime;
  defTime.setDefaultLanguage(lang.c_str());

  if(!setup.gui.style.empty())
    a.setStyle(setup.gui.style.c_str());


  if(!lang.empty()) {

    string qtlang   = "qt_"     +lang;
    string langfile = "tseries_"+lang;

    // translation file for application strings

    for(unsigned int i=0;i<setup.path.lang.size(); i++ )
      if( qt.load( qtlang.c_str(),setup.path.lang[i].c_str()))
        break;

    for(unsigned int i=0;i<setup.path.lang.size(); i++ )
      if( myapp.load( langfile.c_str(),setup.path.lang[i].c_str()))
        break;

    a.installTranslator( &qt    );
    a.installTranslator( &myapp );

  }


  qtsMain *main = new qtsMain(lang, instancename.c_str());

  main->setWindowTitle(title.c_str());
  main->show();

  return a.exec();
}
